### PR TITLE
Introduce functional test for resolve command.

### PR DESCRIPTION
**Pull request by Veebers**
This test ensures that failed hooks are not retried when the argument
'--no-retry' is provided.

## Description of change
As per bug LP:1762979 add a test to ensure the behaviour of "resolve --no-retry" works as expected. (i.e. does not re-try failed hooks).

## Bug reference

https://bugs.launchpad.net/juju/+bug/1762979

### CODE REVIEW

**Comment by Veebers** [comment]
[Add functional test for 'resolve --retry'](https://trello.com/c/kF488eoR/197-add-functional-test-for-resolve-retry)

File: acceptancetests/jujupy/wait_condition.py [hunk 1]
```diff
@@ -321,6 +321,33 @@ def do_raise(self, model_name, status):
         raise AgentsNotStarted(model_name, status)
 
 
+class UnitInstallCondition(BaseCondition):
```
Comment (wallyworld) [inline acceptancetests/jujupy/wait_condition.py]: Ideally these classes would have doc comments to explain what they are for?

File: acceptancetests/jujupy/wait_condition.py [hunk 2]
```diff
@@ -321,6 +321,33 @@ def do_raise(self, model_name, status):
...
+        """Wait until 'current' status and message matches supplied values."""
+        try:
+            unit = status.get_unit(self.unit)
+            unit_status = unit['workload-status']
+            cond_met = unit_status['current'] == self.current \
+                and unit_status['message'] == self.message
+        except KeyError:
+            cond_met = False
+        if not cond_met:
+            yield 'unit-workload', 'not-{}'.format(self.current)
```
Comment (wallyworld) [inline acceptancetests/jujupy/wait_condition.py]: did we want the name of the unit in the message?

File: acceptancetests/jujupy/wait_condition.py [hunk 3]
```diff
@@ -321,6 +321,33 @@ def do_raise(self, model_name, status):
...
+
+    def do_raise(self, model_name, status):
+        raise Exception(
```
Comment (wallyworld) [inline acceptancetests/jujupy/wait_condition.py]: I imagine there will be (or are) other tests which poll status for a particular status/message. Should we introduce a bespoke exception class for that.

File: acceptancetests/assess_resolve.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+import argparse
+import logging
+import sys
+
+from deploy_stack import (
+    BootstrapManager,
+    )
+from jujupy.models import (
+    temporary_model,
+)
```
Comment (wallyworld) [inline acceptancetests/assess_resolve.py]: there's a closing bracket alignment inconsistency here (see above import)
not sure that pylint says is correct

File: acceptancetests/assess_resolve.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    ensure_retry_does_not_rerun_failed_hook(client, local_resolve_charm)
+
+
+def ensure_retry_does_not_rerun_failed_hook(client, resolve_charm):
+        with temporary_model(client, "no-retry") as temp_client:
+            unit_name = 'simple-resolve/0'
+            temp_client.deploy(resolve_charm)
+            temp_client.wait_for(UnitInstallError(unit_name))
+            temp_client.juju('resolve', ('--no-retry', unit_name))
+            temp_client.wait_for(
```
Comment (wallyworld) [inline acceptancetests/assess_resolve.py]: We need to also check show-status-log to ensure the expected hooks have been run,
eg I can't see where we check that the start hook fires when the install hook failure is resolved

Comment (Veebers) [reply acceptancetests/assess_resolve.py]: I missed the new charm in the original push, it's there now. The charm updates the message in the install and start hook script. The UnitInstallActive will wait until it sees the message "'No install hook'" (i.e. start hook fired, the install hook never resolved.)

We could check the show-status-log for the string "juju-unit  executing    running start hook" if we feel we need further confirmation that it ran.

File: acceptancetests/repository/charms/simple-resolve/metadata.yaml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+name: simple-resolve
+maintainer: Christopher Lee <chris.mclachlan-lee@canonical.com>
+summary: Purposfully error on install hook for testing resolve command.
+description: |
+  A testing charm that errors once on install hook then succeeds on the next run.
+categories:
+  - misc
+series:
+  - trusty
```
Comment (wallyworld) [inline acceptancetests/repository/charms/simple-resolve/metadata.yaml]: let's make xenial the default, follow by bionic

File: acceptancetests/repository/charms/simple-resolve/hooks/install [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+fi
+
+# Need to error for x amount of times before juju stops trying automatically.
+run_count=$(cat ${canary_file})
+
+# For now juju is always re-trying. Need to just exit regardless.
+echo $((${run_count}+1)) > ${canary_file}
+exit_error ${run_count}
+
+# if [ ${run_count} -le 3 ]; then
```
Comment (wallyworld) [inline acceptancetests/repository/charms/simple-resolve/hooks/install]: we should leave a comment as to why this code isn't used

File: acceptancetests/assess_resolve.py (hunk 5 above)
Comment (wallyworld) [reply acceptancetests/assess_resolve.py]: I think it's fine, Seeing the charm makes sense. Maybe a code comment saying that the charm sets status etc etc

**Comment by Veebers** [comment]
$$merge$$

**Pull request merged**